cpg-core/src/main/java/de/fraunhofer/aisec/cpg/GraphTransformation.kt

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2022, Fraunhofer AISEC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *                    $$$$$$\  $$$$$$$\   $$$$$$\
+ *                   $$  __$$\ $$  __$$\ $$  __$$\
+ *                   $$ /  \__|$$ |  $$ |$$ /  \__|
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg
+
+import de.fraunhofer.aisec.cpg.frontends.TranslationException
+import de.fraunhofer.aisec.cpg.graph.Node
+import java.lang.Exception
+import java.net.URLEncoder
+import java.util.*
+import java.util.function.Consumer
+
+/**
+ * A generic superclass for components that change the graph. This includes language frontends as the component that does
+ * the initial translation into the CPG-AST, as well as graph enhancing passes. The purpose of this class is to capture all
+ * behavior that we need for this components, such as tracking of handled nodes.
+ */
+abstract class GraphTransformation {
+
+    protected var parserObjectStack = Stack<Any>()
+
+    open fun pushToHandleLog(trackedObject: Any?) {
+        trackedObject?.let { parserObjectStack.push(trackedObject) }
+    }
+
+    open fun popFromHandleLog(trackedObject: Any?) {
+        trackedObject?.let {
+            if (parserObjectStack.contains(trackedObject)) {
+                while (parserObjectStack.pop() !== trackedObject) ;
```

Comment:
```suggestion
                while (parserObjectStack.pop() != trackedObject) ;
```
I would prefer this, as the `.contains()` above also checks with `.equals()`.
Do we even need the `.contains()` above?